### PR TITLE
Upgrade to OkHttp3

### PR DESCRIPTION
**Pull request by AndrewJack**
Update to [OkHttp](https://github.com/square/okhttp) to [OkHttp3](https://publicobject.com/2015/12/12/com-squareup-okhttp3/)

We must also update:
- Fresco to 0.10.0
- okio to 1.8.0

**Motivation**
Reasons for upgrading:
* Issue #4021 
* "We discovered that RN Android sometimes fails to connect to the latest stable version of NGINX when HTTP/2 is enabled. We aren't seeing errors with other HTTP clients so we think it's specific to RN and OkHttp. Square has fixed several HTTP/2 bugs over the past eight months." - @ide
* OkHttp3 will be maintained & improved, but OkHttp2 will only receive [security fixes](https://publicobject.com/2016/02/11/okhttp-certificate-pinning-vulnerability/)
* Cleaner APIs - "Get and Set prefixes are avoided"
* Deprecated/Removed - HttpURLConnection & Apache HTTP 
* React Native apps are currently being forced to bundle two versions of OkHttp (v2 & v3), if another library uses v3
* Improved WebSocket performance - [CHANGELOG.md](https://github.com/square/okhttp/blob/master/CHANGELOG.md#version-310)
* The creators [recommend](https://publicobject.com/2016/02/11/okhttp-certificate-pinning-vulnerability/) keeping OkHttp up to date.

 > If you’re using OkHttp, you should upgrade to the latest version immediately! Staying up to date on OkHttp is a good idea – we track the latest HTTPS cipher suites & TLS versions to balance connectivity with security. It’s like keeping the browser up to date on your computer: staying current is the best way to stay safe.

See the [CHANGELOG.md](https://github.com/square/okhttp/blob/master/CHANGELOG.md) for more info.

**Test plan**
- Circle Ci
- Build a project from source of this branch - https://github.com/AndrewJack/react-native-test/tree/okhttp3_6113

**Still to do:**
- [x] Use new CookieJar class

### CODE REVIEW

**Comment by AndrewJack** [comment]
Depends on - https://github.com/facebook/fresco/pull/875

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 1]
```diff
@@ -97,7 +99,7 @@ public NetworkingModule(ReactApplicationContext reactContext, OkHttpClient clien
 
   @Override
   public void initialize() {
-    mClient.setCookieHandler(mCookieHandler);
+    // mClient.setCookieHandler(mCookieHandler); // TODO
```
Comment (AndrewJack) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: setCookieHandler has been removed. the new api is called CookieJar which is added when the OkHttpClient is built

**Comment by bestander** [comment]
Hey Andrew.
This is a noble thing you are doing here :)

Just a heads up that this task is even harder than you may think because internally at FB we have all projects in one codebase and all thirdparty dependencies are shared.

Projects that use OkHttp2 will become runtime incompatible with ReactNative and we may need to update all libs to OkHttp3 at once.

**Comment by AndrewJack** [comment]
@bestander 
After working on this, I can see that this task isn't as easy as I first thought. 

> Projects that use OkHttp2 will become runtime incompatible with ReactNative and we may need to update all apps and libs to use OkHttp3 at once.

The package of OkHttp 3 is different to OkHttp 2. So they should work along side each while teams & projects migrate.
v2 `com.squareup.okhttp` 
v3 `okhttp3`

Would upgrading to 2.7.4 cause similar issues with FB's internal codebase?

**Comment by bestander** [comment]
> The package of OkHttp 3 "okhttp3" is different to OkHttp 2 "com.squareup.okhttp". So they should work along side each while teams & projects migrate.

Imagine you have a large app that uses OkHttp2 and React Native.
And React Native that uses OkHttp3.
In best case scenario the large app will grow in size because it will include both OkHttp2 and OkHttp3 jars. This is not good because at FB we are optimising to have smaller apps.
In the worst case scenario OkHttp2 and OkHttp3 may share some classes and be incompatible internally or maybe use common state and behave unexpectedly.

2.7.4 update should be easier if the API did not break.
But there needs to be a benefit to do that.

**Comment by AndrewJack** [comment]
Maybe this is just a bit premature. 

I can imagine that scenario switching in a few months time. Where we will have apps using OkHttp3 but react native still uses OkHttp2 thus still bloating the app.

**Comment by bestander** [comment]
Yeah, I understand the problem.
I hope you did not waste your time on that and we still would be able to upgrade eventually.

I'll raise the question in the internal engineering group.

**Comment by AndrewJack** [comment]
Thanks. 

Shall we close this PR? It still needs some work anyway and there is already a discussion on #4021.

**Comment by bestander** [comment]
@AndrewJack yes, I think it would be better to close for now.
By the time we can merge this PR a rebase will be required and code will have plenty of conflicts.

There is no particular reason to not to upgrade, just the benefit of update must outweigh the effort and risks to update.
Maybe if you could provide a backwards compatibility layer so that FB could keep using v2 and the OSS community v3 at the same time then it could be an acceptable solution.

Anyway, this would need a broader discussion, maybe on the OSS contributors board, to get the opinions from all interested parties.
Thanks for the contribution anyway, please don't be discouraged by this case.

**Comment by ide** [comment]
Now that Fresco's using OkHttp 3 can we update RN as well? There are bugs in OkHttp 2 when connecting to NGINX (https://www.facebook.com/groups/reactnativeoss/permalink/1549975525299076/) and the JARs for OkHttp 3 + Okio are pretty small (about 400KB) so I'm thinking the pros outweigh the cons now.

**Comment by ide** [comment]
@AndrewJack based on that facebook thread I linked above, it sounds upgrading to OkHttp 3 is worth investigating again. Fresco has been updated as well. 

The RN team at FB can help with code reviews but is too busy to do the implementation. Would you be interested in re-opening and updating this PR?

**Comment by AndrewJack** [comment]
@ide I'll try to find some time to have another go at this.

**Comment by bestander** [comment]
I've seen the comment, will get back to this

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java [hunk 2]
```diff
@@ -137,10 +137,10 @@ public String getName() {
   @Override
   public void onCatalystInstanceDestroy() {
     mShuttingDown = true;
-    mClient.cancel(null);
+    OkHttpCallUtil.cancelTag(mClient, null);
 
     mCookieHandler.destroy();
-    mClient.setCookieHandler(null);
+    // mClient.setCookieHandler(null); // TODO
```
Comment (AndrewJack) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: TODO - handle cookies

**Comment by facebook-github-bot-9** [comment]
@AndrewJack updated the pull request.

File: ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java (hunk 1 above)
Comment (AndrewJack) [inline ReactAndroid/src/main/java/com/facebook/react/modules/network/NetworkingModule.java]: TODO: Need to handle cookies

File: ReactAndroid/src/main/java/com/facebook/react/common/OkHttpCallUtil.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        call.cancel();
+      }
+    }
+    for (Call call : client.dispatcher().runningCalls()) {
+      if (tag.equals(call.request().tag())) {
+        call.cancel();
+      }
+    }
+  }
+}
```
Comment (AndrewJack) [inline ReactAndroid/src/main/java/com/facebook/react/common/OkHttpCallUtil.java]: Any guidance on where this should go?

**Comment by facebook-github-bot-4** [comment]
@AndrewJack updated the pull request.

File: ReactAndroid/src/main/java/com/facebook/react/common/OkHttpCallUtil.java (hunk 3 above)
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/common/OkHttpCallUtil.java]: ReactAndroid/src/main/java/com/facebook/react/modules/network looks like the best place

File: ReactAndroid/src/main/java/com/facebook/react/bridge/JSDebuggerWebSocketClient.java [hunk 4]
```diff
@@ -59,11 +60,11 @@ public void connect(String url, JSDebuggerCallback callback) {
...
-    mHttpClient = new OkHttpClient();
-    mHttpClient.setConnectTimeout(10, TimeUnit.SECONDS);
-    mHttpClient.setWriteTimeout(10, TimeUnit.SECONDS);
-    // Disable timeouts for read
-    mHttpClient.setReadTimeout(0, TimeUnit.MINUTES);
+    mHttpClient = new OkHttpClient.Builder()
+      .connectTimeout(10, TimeUnit.SECONDS)
+      .writeTimeout(10, TimeUnit.SECONDS)
+      .readTimeout(0, TimeUnit.MINUTES)
+      .build(); // Disable timeouts for read
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/JSDebuggerWebSocketClient.java]: Move this comment next to or above the .readTimeout() call

File: ReactAndroid/src/main/java/com/facebook/react/common/OkHttpCallUtil.java (hunk 3 above)
Comment (AndrewJack) [inline ReactAndroid/src/main/java/com/facebook/react/common/OkHttpCallUtil.java]: I got a Buck cycle found error when I put it there initially.

**Comment by AndrewJack** [comment]
Started a Cookie implementation - a backward compatible one. Could probably make a better implementation with the new APIs.
https://github.com/AndrewJack/react-native/commit/f2517f04d27c3f87c28d59b3fdcb8b0e9256cd96

**Comment by bestander** [comment]
@AndrewJack wow, that is a big and useful one, thanks for doing this.
Thanks for getting back to this.
Ping me once the review is done and I'll import it, a few links in internal FB codebase need to be fixed.

**Comment by facebook-github-bot-7** [comment]
@bestander would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

File: ReactAndroid/src/main/java/com/facebook/react/common/OkHttpCallUtil.java (hunk 3 above)
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/common/OkHttpCallUtil.java]: OK, make a new directory in facebook/react/common/network then.

File: ReactAndroid/src/main/java/com/facebook/react/bridge/JSDebuggerWebSocketClient.java [hunk 5]
```diff
@@ -187,18 +186,17 @@ private void triggerRequestSuccess(int requestID, @Nullable String response) {
     }
   }
 
-  @Override
-  public void onMessage(BufferedSource payload, WebSocket.PayloadType type) throws IOException {
-    if (type != WebSocket.PayloadType.TEXT) {
-      FLog.w(TAG, "Websocket received unexpected message with payload of type " + type);
+  public void onMessage(ResponseBody response) throws IOException {
```
Comment (ide) [inline ReactAndroid/src/main/java/com/facebook/react/bridge/JSDebuggerWebSocketClient.java]: @Override?

**Comment by ide** [comment]
Seems good -- @AndrewJack can you add your cookiejar commit to this PR? Generally smaller PRs are better but in this case we want to atomically upgrade all of the networking code at once.

**Comment by AndrewJack** [comment]
Yep sure I just pushed it there quickly until I had time to test it out properly.

**Comment by facebook-github-bot-9** [comment]
@AndrewJack updated the pull request.

**Comment by AndrewJack** [comment]
Trying to test this out, but keep running into this issue (https://github.com/facebook/react-native/issues/7336) has anyone seen this before?

**Comment by facebook-github-bot-2** [comment]
@bestander would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Comment by facebook-github-bot-6** [comment]
@AndrewJack updated the pull request.

**Comment by AndrewJack** [comment]
@bestander It's ready to be reviewed. 

It's probably worth testing on a few real world projects before this is pushed out. I've tested on this simple project - https://github.com/AndrewJack/react-native-test/tree/okhttp3_6113

**Comment by bestander** [comment]
Thanks, Andrew.
This will take a few days to test, I'll need to change a few links and run
a code search.

**Comment by bestander** [comment]
@facebook-github-bot import

**Comment by facebook-github-bot-2** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/536127679901257/int_phab) to review.

**Comment by facebook-github-bot-5** [comment]
@bestander would you mind taking a look at this pull request? It's been a while since the last commit was reviewed.

**Pull request closed without merging**